File: ios/Support/Metrics.swift
import QuartzCore

/** Counters only. Aggregated and sent every 2 s when diagnostics are on. */
final class SessionMetrics {
  var errors = 0
  var droppedFrames = 0
  private var startups: [Int] = []
  private var firstFrames = 0
  private var rebufferCount = 0
  private var rebufferMs = 0.0
  private var bufferStartedAt: CFTimeInterval?

  func recordStartup(_ ms: Int, fromCache: Bool) {
    firstFrames += 1
    startups.append(ms)
    if startups.count > 200 { startups.removeFirst(startups.count - 200) }
  }

  func bufferStart() {
    guard bufferStartedAt == nil else { return }
    bufferStartedAt = CACurrentMediaTime()
    rebufferCount += 1
  }

  func bufferEnd() {
    guard let start = bufferStartedAt else { return }
    rebufferMs += (CACurrentMediaTime() - start) * 1000
    bufferStartedAt = nil
  }

  func payload(cache: VideoCache.Stats, playersActive: Int, playersPrepared: Int, playersTotal: Int,
               liveObservers: Int, liveSurfaces: Int, memoryWarnings: Int) -> [String: Any] {
    let sorted = startups.sorted()
    func pct(_ p: Double) -> Int { sorted.isEmpty ? 0 : sorted[min(sorted.count - 1, Int(Double(sorted.count) * p))] }
    return [
      "startupMsP50": pct(0.5), "startupMsP90": pct(0.9), "firstFrames": firstFrames,
      "rebufferCount": rebufferCount, "rebufferMs": Int(rebufferMs),
      "cacheHits": cache.hits, "cacheMisses": cache.misses,
      "bytesActive": cache.bytesActive, "bytesSpeculative": cache.bytesSpeculative, "bytesWasted": cache.bytesWasted,
      "preloadCancels": cache.preloadCancels, "droppedFrames": droppedFrames,
      "playersActive": playersActive, "playersPrepared": playersPrepared, "playersTotal": playersTotal,
      "liveObservers": liveObservers, "liveSurfaces": liveSurfaces, "memoryWarnings": memoryWarnings,
      "errors": errors, "videoCacheBytes": 0, "imageCacheBytes": 0,
    ]
  }
}

File: android/src/main/java/expo/modules/infinitemedia/Metrics.kt
package expo.modules.infinitemedia

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/** Counters for diagnostics. Aggregated here, emitted as one map every 2 s when enabled. */
internal class Metrics {
  private val startups = ArrayList<Long>()
  var rebufferCount = 0
  var rebufferMs = 0L
  var cacheHits = 0
  var cacheMisses = 0
  var preloadCancels = 0
  var droppedFrames = 0L
  var errors = 0

  // Written from Media3 loader threads.
  val bytesActive = AtomicLong()
  val bytesSpeculative = AtomicLong()
  /** Speculative bytes per cache key, cleared once the key actually plays. */
  val speculativeByKey = ConcurrentHashMap<String, AtomicLong>()

  fun startup(ms: Long) {
    startups.add(ms)
    // Bounded, a long session only needs recent history for percentiles.
    if (startups.size > 200) startups.removeAt(0)
  }

  fun played(key: String) {
    speculativeByKey.remove(key)
  }

  fun snapshot(extra: Map<String, Any>, wastedKeys: Set<String>): Map<String, Any> {
    val sorted = startups.sorted()
    fun pct(p: Double) = if (sorted.isEmpty()) 0L else sorted[((sorted.size - 1) * p).toInt()]
    val wasted = wastedKeys.sumOf { speculativeByKey[it]?.get() ?: 0L }
    return mapOf(
      "startupMsP50" to pct(0.5),
      "startupMsP90" to pct(0.9),
      "firstFrames" to startups.size,
      "rebufferCount" to rebufferCount,
      "rebufferMs" to rebufferMs,
      "cacheHits" to cacheHits,
      "cacheMisses" to cacheMisses,
      "bytesActive" to bytesActive.get(),
      "bytesSpeculative" to bytesSpeculative.get(),
      "bytesWasted" to wasted,
      "preloadCancels" to preloadCancels,
      "droppedFrames" to droppedFrames,
      "errors" to errors
    ) + extra
  }
}
